pkg/planner/core/plan_cache.go

```diff
@@ -104,8 +104,32 @@ func planCachePreprocess(ctx context.Context, sctx sessionctx.Context, isNonPrep
 		return errors.Trace(err)
 	}
 
+	schemaNotMatch := false
+	for i := 0; i < len(stmt.dbName); i++ {
+		_, ok := is.TableByID(stmt.tbls[i].Meta().ID)
+		if !ok {
+			tblByName, err := is.TableByName(stmt.dbName[i], stmt.tbls[i].Meta().Name)
+			if err != nil {
+				return plannererrors.ErrSchemaChanged.GenWithStack("Schema change caused error: %s", err.Error())
+			}
+			delete(stmtAst.RelateVersion, stmt.tbls[i].Meta().ID)
+			stmt.tbls[i] = tblByName
+			stmtAst.RelateVersion[tblByName.Meta().ID] = tblByName.Meta().TableVersion
+		}
+		newTbl, err := tryLockMDLAndUpdateSchemaIfNecessary(sctx.GetPlanCtx(), stmt.dbName[i], stmt.tbls[i], is)
+		if err != nil {
+			schemaNotMatch = true
+			continue
+		}
+		if stmt.tbls[i].Meta().TableVersion != newTbl.Meta().TableVersion {
+			schemaNotMatch = true
+		}
+		stmt.tbls[i] = newTbl
+		stmtAst.RelateVersion[newTbl.Meta().ID] = newTbl.Meta().TableVersion
+	}
```

Comment:
This part doesn't belong to step 2 or step 3. Please split it to a separate step. And add comments for this part.